app/services/api/declarations/query.rb

```diff
@@ -0,0 +1,130 @@
+module API::Declarations
+  class Query
+    include Queries::FilterIgnorable
+
+    attr_reader :scope, :lead_provider_id
+
+    def initialize(
+      lead_provider_id: :ignore,
+      contract_period_years: :ignore,
+      teacher_api_ids: :ignore,
+      delivery_partner_api_ids: :ignore,
+      updated_since: :ignore
+    )
+      @lead_provider_id = lead_provider_id
+      @scope = Declaration.distinct
+
+      where_lead_provider_is(lead_provider_id)
+      where_contract_period_year_in(contract_period_years)
+      where_teacher_is(teacher_api_ids)
+      where_delivery_partner_is(delivery_partner_api_ids)
+      where_updated_since(updated_since)
+    end
+
+    def declarations
+      preload_associations(block_given? ? yield(scope) : scope)
+    end
+
+    def declaration_by_api_id(api_id)
+      return preload_associations(scope).find_by!(api_id:) if api_id.present?
+
+      fail(ArgumentError, "api_id needed")
+    end
+
+    def declaration_by_id(id)
+      return preload_associations(scope).find(id) if id.present?
+
+      fail(ArgumentError, "id needed")
+    end
+
+  private
+
+    def preload_associations(results)
+      results
+        .strict_loading
+        .includes(
+          :payment_statement,
+          :clawback_statement,
+          training_period: [
+            { ect_at_school_period: :teacher },
+            { mentor_at_school_period: :teacher },
+            :delivery_partner,
+            :lead_provider,
+          ]
+        )
+    end
+
+    def where_lead_provider_is(lead_provider_id)
+      return if ignore?(filter: lead_provider_id)
+
+      # Direct declarations: training period belongs to this lead provider
+      direct_scope = scope
+        .joins(training_period: :lead_provider)
+        .where(lead_providers: { id: lead_provider_id })
+
+      # Previous declarations: teachers who trained with this LP (with date/status constraints)
+      previous_scope = previous_declarations_for_lead_provider(lead_provider_id)
+
+      @scope = scope.where(id: direct_scope.select(:id)).or(scope.where(id: previous_scope.select(:id)))
+    end
+
+    def previous_declarations_for_lead_provider(lead_provider_id)
```

Comment:
Hi @mooktakim just an idea.. not sure tho.. could we use a new metadata table for that?